File: tests/test_threads.py
import pytest

pytestmark = pytest.requires_threading

with pytest.suppress(ImportError):
    import threading
    from pybind11_tests import threads as t


def test_threads():
    def pythread_routine():
        threading.current_thread()._return = t.check_pythread()

    thread = threading.Thread(target=pythread_routine)
    thread.start()
    thread.join()
    assert thread._return

    assert t.check_cthread()

File: tests/test_threads.cpp
#include "pybind11_tests.h"

#if defined(WITH_THREAD)

#include <thread>

static bool check_threadstate() {
    return PyGILState_GetThisThreadState() == PyThreadState_Get();
}

TEST_SUBMODULE(threads, m) {
    m.def("check_pythread", []() -> bool {
        py::gil_scoped_acquire acquire1;
        return check_threadstate();
    }, py::call_guard<py::gil_scoped_release>())
    .def("check_cthread", []() -> bool {
        bool result = false;
        std::thread thread([&result]() {
            py::gil_scoped_acquire acquire;
            result = check_threadstate();
        });
        thread.join();
        return result;
    }, py::call_guard<py::gil_scoped_release>());
}

#endif
